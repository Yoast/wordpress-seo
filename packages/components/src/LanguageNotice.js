/* External dependencies */
import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { __, sprintf } from "@wordpress/i18n";
import interpolateComponents from "interpolate-components";

/* Internal dependencies */
import { makeOutboundLink } from "@yoast/helpers";

const YoastLanguageNotice = styled.p`
	margin: 1em 0;
`;

const ChangeLanguageLink = makeOutboundLink( styled.a`
	margin-left: 4px;
` );

/**
 * Returns the LanguageNotice component.
 *
 * @param {Object} props                    The props for this language notice wrapper.
 * @param {string} props.changeLanguageLink The URL where the language can be changed.
 * @param {bool}   props.canChangeLanguage  Whether or not the language can be changed.
 * @param {string} props.language           The currently set language.
 * @param {bool}   props.showLanguageNotice Whether or not the language notice is shown.
 *
 * @returns {ReactElement} The LanguageNotice component.
 */
export default class LanguageNotice extends PureComponent {
	/**
	 * Renders the language notice. Provides a link to a setting page if the currently set language can be changed.
	 * Provides a message to contact an administrator otherwise.
	 *
	 * @returns {ReactElement} The rendered language notice.
	 */
	render() {
		const {
			changeLanguageLink,
			canChangeLanguage,
			language,
			showLanguageNotice,
		} = this.props;

		if ( ! showLanguageNotice ) {
			return null;
		}

		// Determine the correct text.
		/* Translators: %s expands to the actual language. */
		let text = __( "Your site language is set to %s. ", "wordpress-seo" );
		if ( ! canChangeLanguage ) {
			/* Translators: %s expands to the actual language. */
			text = __( "Your site language is set to %s. If this is not correct, contact your site administrator.", "wordpress-seo" );
		}

		// Replace the %s with a strong marked language.
		text = sprintf( text, `{{strong}}${ language }{{/strong}}` );

		// Replace the strong marking with an actual ReactComponent.
		text = interpolateComponents( {
			mixedString: text,
			components: { strong: <strong /> },
		} );

		return (
			<YoastLanguageNotice>
				{ text }
				{ canChangeLanguage && <ChangeLanguageLink href={ changeLanguageLink }>
					{ __( "Change language", "wordpress-seo" ) }
				</ChangeLanguageLink> }
			</YoastLanguageNotice>
		);
	}
}

export const languageNoticePropType = {
	changeLanguageLink: PropTypes.string.isRequired,
	canChangeLanguage: PropTypes.bool,
	language: PropTypes.string.isRequired,
	showLanguageNotice: PropTypes.bool,
};

LanguageNotice.propTypes = languageNoticePropType;

LanguageNotice.defaultProps = {
	canChangeLanguage: false,
	showLanguageNotice: false,
};
