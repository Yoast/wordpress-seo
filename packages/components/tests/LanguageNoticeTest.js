/* External dependencies */
import React from "react";
import renderer from "react-test-renderer";

/* Internal dependencies */
import LanguageNotice from "../src/LanguageNotice";

describe( "LanguageNotice", () => {
	it( "matches the snapshot by default", () => {
		const component = renderer.create(
			<LanguageNotice
				changeLanguageLink="#"
				language="English"
			/>
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "matches the snapshot when showLanguageNotice equals true", () => {
		const component = renderer.create(
			<LanguageNotice
				changeLanguageLink="#"
				language="English"
				showLanguageNotice={ true }
			/>
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "matches the snapshot with a changed language link", () => {
		const component = renderer.create(
			<LanguageNotice
				changeLanguageLink="http://www.example.com"
				canChangeLanguage={ true }
				language="English"
				showLanguageNotice={ true }
			/>
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "matches the snapshot with another language", () => {
		const component = renderer.create(
			<LanguageNotice
				changeLanguageLink="http://www.example.com"
				canChangeLanguage={ true }
				language="Nederlands"
				showLanguageNotice={ true }
			/>
		);

		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );
} );
