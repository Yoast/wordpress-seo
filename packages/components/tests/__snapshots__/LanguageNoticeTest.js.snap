// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageNotice matches the snapshot by default 1`] = `null`;

exports[`LanguageNotice matches the snapshot when showLanguageNotice equals true 1`] = `
.c0 {
  margin: 1em 0;
}

<p
  className="c0"
>
  Your site language is set to 
  <strong>
    English
  </strong>
  . If this is not correct, contact your site administrator.
</p>
`;

exports[`LanguageNotice matches the snapshot with a changed language link 1`] = `
.c2 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 1em 0;
}

.c1 {
  margin-left: 4px;
}

<p
  className="c0"
>
  Your site language is set to 
  <strong>
    English
  </strong>
  . 
  <a
    className="c1"
    href="http://www.example.com"
    rel="noopener"
    target="_blank"
  >
    Change language
    <span
      className="c2"
    >
      (Opens in a new browser tab)
    </span>
  </a>
</p>
`;

exports[`LanguageNotice matches the snapshot with another language 1`] = `
.c2 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 1em 0;
}

.c1 {
  margin-left: 4px;
}

<p
  className="c0"
>
  Your site language is set to 
  <strong>
    Nederlands
  </strong>
  . 
  <a
    className="c1"
    href="http://www.example.com"
    rel="noopener"
    target="_blank"
  >
    Change language
    <span
      className="c2"
    >
      (Opens in a new browser tab)
    </span>
  </a>
</p>
`;
